rfcs/0036-hard-fork-disaster-recovery.md

```diff
@@ -0,0 +1,235 @@
+    and provide access to the new binary
+
+To remedy the problems that led to a failure, the Coda software will
+likely change in some significant way when the network is
+restarted. Using a new protocol version, with a new major/minor or minor
+version, will require node operators to upgrade their software.
+
+CLI command to save state
+-------------------------
+
+The Coda developers will choose a node with a root to represent the
+starting point of the hard fork. The choice of node is beyond the scope of
+this RFC.
+
+The CLI command can be in the `internal` group of commands, since
+it's meant for use in extraordinary circumstances. A suggested
+name is `save-hard-fork-data`. That command communicates with the
+running node daemon via the daemon-RPC mechanism used in other
+client commands.
+
+Let `frontier` be the current transition frontier. When the CLI command
+is run, the daemon saves the following data:
+
+ - its root
+
+   this is an instance of `Protocol_state.value`, retrievable via
+
+	 ```ocaml
+      let full = Transition_frontier.full_frontier frontier in
+      let root = full.root in
+	  root |> find_protocol_state
+     ```
+
+ - the SNARK proof for that root
+
+   this is an instance of `Proof.t`, retrievable via
+
+   ```ocaml
+      let full = Transition_frontier.full_frontier frontier in
+      let transition_with_hash,_ = Full_frontier.(root_data full).transition in
+	  let transition = With_hash.data transition_with_hash in
+	  transition.protocol_state_proof
+   ```
+
+ - the SNARKed ledger corresponding to the root
+
+   this is an instance of `Coda_base.Ledger.Any_ledger.witness`, retrievable
+    via
+
+   ```ocaml
+    let full = Transition_frontier.full_frontier frontier in
+    full.root_ledger
+   ```
+   Note: There appears to be a mechanism in `Persistent_root` for saving the
+   root ledger, but it appears only to store a ledger hash, and not the ledger itself.
+
+ - two epoch ledgers
+
+   there is pending PR #4115 which allows saving epoch ledgers to RocksDB databases
+
+   which two epoch ledgers needed depends on whether the root is in the epoch current
+     at the time of the network pause, or in the previous one:
+
+	 - if the root is in the current epoch, the two ledgers needed are
+	    `staking_epoch_snapshot` and `next_epoch_snapshot`, as in the PR
+     - if the root is in the previous epoch, the two ledger needed are
+        `staking_epoch_snapshot` and `previous_epoch_snapshot` (not implemented
+	    in the PR)
+
+ - the breadcrumb at the root
+
+   this is an instance of `Breadcrumb.t`, retrievable via
+
+   ```ocaml
+     let full = Transition_frontier.full_frontier frontier in
+	 let root = full.root in
+     Full_frontier.find_exn full root
+   ```
+   The breadcrumb contains a validated block, and a staged ledger,
+   which contains a scan state.
+
+   As of this writing, there's no serialization code to fully capture a
+   breadcrumb; that would have to be written.
+
```

Comment:
Are you suggesting constructing a staged ledger in that same way, and not using the root breadcrumb?

The serialization code for `Breadcrumb.t` might be as simple as adding a deriver to the type.

Reply:
Yeah, we wouldn't to serialize another ledger

Reply:
Do you mean, we wouldn't have to add serialization code, or, there's an already-serialized ledger that exists on-disk that we can use?

Reply:
I mean we don't have to serialize the ledger in staged ledger because we can construct it using the snarked ledger (written to the disk using rocks db) and the scan state. So instead of serializing the entire breadcrumb we only serialize the required parts

Reply:
I see, thanks.